### ISSUE TITLE
scalapb_proto_library additional scala sources

### ISSUE THREAD

**Issue by holinov** [bazelbuild/rules_scala#753]
Now it is impossible to add some custom scala sources into scalapb library. So i could not do something like this:
```scala
trait TagRecordCompanionTrait {
    implicit val ordering: Ordering[TagRecord] = (x: TagRecord, y: TagRecord) => {
     if (x.count < y.count) 1
     else if (x.count > y.count) -1
     else x.name compareTo y.name
   }
}
```
```proto
message TagRecord {
    option (scalapb.message).companion_extends = "TagRecordCompanionTrait";

    required string name = 1;
    required int64 count = 2;
}
```
The only way to do this is to use preamble in proto file

**Comment by ianoc** [comment]
This appears to be a duplicate of https://github.com/bazelbuild/rules_scala/issues/751 

If you want to weigh in there around the discussion that would be great

**Comment by holinov** [comment]
No. That issue is about adding some external dep. In my case i need to add some scala sources into scalapb_library. I need to create base trait for message/message companion and set sef-type to message (so i could not compile my scala files and flies from protobuf in same scala library because of cyclic dependency)

**Comment by ianoc** [comment]
Interesting, scalapb and its many options around this really make the aspect like stuff quite tricky. 
OOC have you managed to use any of the scalapb rules since/if so any change/improvement in performance? (it was huge for us, but we probably have more proto's than most).


On this, if you have any thoughts on how/what an API should look like here or how generic/often you need to do this would love to hear it. Some thoughts of ways we could try do it:

1) Use the experimental (new) api around adding another printer and emit the scala to do ordering or other things dynamically from that. (this is on master today so it would be usable -- just needs more docs written, and as more people use it we may update the API's to bake it in more too).
2) Use a manual scalapb rule and not use the aspects at all. We could just share some of the generation steps but have a non-aspect set of rules here too. Mixing the two would be odd in terms of duplicate classes. But if folks aren't getting a win and it makes stuff too complex then thats maybe not worth forcing everything down one path here.
3) It feels maybe a bit iffy of a plot. But we could specify in the toolchain a per target filegroup to inject similar to the dependencies discussed in #751. The two could probably be designed to compliment each other reasonably well

**Comment by holinov** [comment]
I use this option all most every time i use scalapb. (it allows to extend generated messages/companions with methods and implicits)
When i'm thinking about api i think of it as 
```bazel

proto_library(
    name = "common_proto",
    srcs = [ "common.proto" ],
    visibility = ["//visibility:public"],
    proto_source_root = "proto",
)

proto_library(
    name = "tags_proto",
    deps = [
        ":common_proto",
        "@com_google_protobuf//:any_proto"
    ],
    srcs = [ "tags.proto" ],
    visibility = ["//visibility:public"],
)

scalapb_proto_library(
    name = "proto_models",
    visibility = ["//visibility:public"],
    srcs = [ "BaseTraits.scala" ],
    deps = [
        ":protobuff_commons",
        ":tags_proto",
    ],
)
```
So `srcs` in `scalapb_proto_library` just appends all files passed to it to protobuf-generated sources and compiles them together

**Comment by ianoc** [comment]
That sort of API would block the use of aspects, since aspects do their own shadow graph of compiling/building of things. So that sounds like you'd favor (2) here, manual protobuf-scala libraries than the aspect. 

@ittaiz @simuons I know Wix had some issues/concerns with the aspects stuff too, not sure if these are the similar sorts of challenges. Would love if you guys have any thoughts on this ^^ to weigh in too

**Comment by holinov** [comment]
maybe you could point me at some kind ow workaround?
i'm eager to switch to bazel and this issue is only thing that blocks me.
maybe you could show me some ''legacy" code that i could use to implement such rule? (i'm totaly new to bazel)

**Comment by ittaiz** [comment]
@holinov what we did is copy over some of the code before the aspects [change](https://github.com/bazelbuild/rules_scala/commit/74192b25112bbd8de7051e8834435825ff819bf3).
Take a look at the diff to see the previous rule.
This can get you unstuck.
On #763 we're trying to think how to proceed in a more structured form

**Comment by luser** [comment]
I did a little work to try out bazel at my current job and Im stuck on this issue presently. We've got `scalapb.field` references in our protos to types from other .scala files, so we need to compile them together. Unfortunately since we're also using google's `{wrapper,descriptor}.proto` and `scalapb.proto` I can't define a toolchain that includes our scala sources by default because the scala sources also depend on importing the generated types from the google/scalapb protos, so we'd wind up with a circular dependency. (Unless there's some way to force the use of a toolchain for a particular `scalapb_proto_library` rule that I'm not aware of.)
